### PR TITLE
Fix In-place Assignments for `PiecewiseRationalQuadratic` Compatibility with `functorch` and `torch2.0`

### PR DESCRIPTION
**Pull request by HamidrezaKmK**
Hello,

While working on my project ([OOD Detection using Manifolds](https://github.com/HamidrezaKmK/ood-detection)), I noticed an issue with the current implementation of the `PiecewiseRationalQuadratic` coupling layers. The in-place masked assignments, specifically:

```python
outputs[outside_interval_mask] = inputs[outside_interval_mask]
```
pose challenges when constructing the computation graph in both `functorch` and `torch2.0`.

To address this, I've made necessary modifications in my fork, ensuring a functional adaptation that aligns with these libraries without altering the primary functionality of the layer.

I kindly ask for your review of these changes. If they align with your vision and maintain the library's integrity, I'd appreciate their incorporation into the main branch.

Thank you for your time and consideration.

Best,
Hamid

### CODE REVIEW

File: nflows/transforms/splines/rational_quadratic.py [hunk 1]
```diff
@@ -21,7 +21,6 @@ def unconstrained_rational_quadratic_spline(
     min_bin_width=DEFAULT_MIN_BIN_WIDTH,
     min_bin_height=DEFAULT_MIN_BIN_HEIGHT,
     min_derivative=DEFAULT_MIN_DERIVATIVE,
-    enable_identity_init=False,
```
Comment (arturbekasov) [inline nflows/transforms/splines/rational_quadratic.py:24]: Is this on purpose, or we should re-base the changes?

File: nflows/transforms/splines/rational_quadratic.py [hunk 2]
```diff
@@ -34,16 +33,18 @@ def unconstrained_rational_quadratic_spline(
         constant = np.log(np.exp(1 - min_derivative) - 1)
         unnormalized_derivatives[..., 0] = constant
         unnormalized_derivatives[..., -1] = constant
-
-        outputs[outside_interval_mask] = inputs[outside_interval_mask]
-        logabsdet[outside_interval_mask] = 0
+        
+        outputs = torch.where(outside_interval_mask, inputs, outputs)
```
Comment (arturbekasov) [inline nflows/transforms/splines/rational_quadratic.py:37]: Should we drop the zero init [above](https://github.com/bayesiains/nflows/pull/77/commits/04989c4a65c70c6cda29cf8c645140f98f383b59#diff-56a605777531dad36582c6a82ec8a384705f36c6c342215093586bdfadd58198R28-R29) if we're copying here anyway? I.e. `outputs = torch.where(outside_interval_mask, inputs, torch.zeros_like(inputs))`.

File: nflows/transforms/splines/rational_quadratic.py [hunk 3]
```diff
@@ -34,16 +33,18 @@ def unconstrained_rational_quadratic_spline(
...
+        logabsdet = torch.where(outside_interval_mask, torch.zeros_like(logabsdet), logabsdet)
```
Comment (arturbekasov) [inline nflows/transforms/splines/rational_quadratic.py:38]: Actually, does the original line even have an effect? We're assigning zeros to what is already initialized to zeros.

File: nflows/transforms/splines/rational_quadratic.py [hunk 4]
```diff
@@ -34,16 +33,18 @@ def unconstrained_rational_quadratic_spline(
...
     else:
         raise RuntimeError("{} tails are not implemented.".format(tails))
 
     if torch.any(inside_interval_mask):
         (
-            outputs[inside_interval_mask],
-            logabsdet[inside_interval_mask],
+            # outputs[inside_interval_mask],
```
Comment (arturbekasov) [inline nflows/transforms/splines/rational_quadratic.py:44]: Let's not leave unused code in comments.

File: nflows/transforms/splines/rational_quadratic.py [hunk 5]
```diff
@@ -97,11 +114,7 @@ def rational_quadratic_spline(
     cumwidths[..., -1] = right
     widths = cumwidths[..., 1:] - cumwidths[..., :-1]
 
-    if enable_identity_init: #flow is the identity if initialized with parameters equal to zero
```
Comment (arturbekasov) [inline nflows/transforms/splines/rational_quadratic.py:100]: Same as earlier comment: we want to keep this, not sure if this is on purpose.

File: nflows/transforms/splines/rational_quadratic.py [hunk 6]
```diff
@@ -57,12 +58,29 @@ def unconstrained_rational_quadratic_spline(
             min_bin_width=min_bin_width,
             min_bin_height=min_bin_height,
             min_derivative=min_derivative,
-            enable_identity_init=enable_identity_init,
         )
+        
+        
+        # turn inside_interval_mask into an int tensor
```
Comment (arturbekasov) [inline nflows/transforms/splines/rational_quadratic.py:64]: Could we simplify below by using [masked_scatter](https://pytorch.org/docs/stable/generated/torch.Tensor.masked_scatter.html)?

**Review by arturbekasov — CHANGES_REQUESTED** [review]
Hey Hamid,

Thanks for taking the time for submit the PR. Left a few comments.

I am not against making the code more functional, but I would like to be a bit careful about performance implications of the additional copies. Is there a chance to see some before/after performance stats of the function?

Cheers,

Artur
